### PR TITLE
implement the new handshake

### PR DESCRIPTION
**Pull request by marten-seemann**
Please read https://github.com/libp2p/specs/pull/151.

### CODE REVIEW

File: crypto.go [hunk 1]
```diff
@@ -95,70 +102,95 @@ func getRemotePubKey(chain []*x509.Certificate) (ic.PubKey, error) {
...
+	}
+	var sk signedKey
+	if _, err := asn1.Unmarshal(keyExt.Value, &sk); err != nil {
+		return nil, fmt.Errorf("unmarshalling signed certificate failed: %s", err)
+	}
+	pubKey, err := crypto.UnmarshalPublicKey(sk.PubKey)
+	if err != nil {
+		return nil, fmt.Errorf("unmarshalling public key failed: %s", err)
+	}
+	certKeyPub, err := x509.MarshalPKIXPublicKey(cert.PublicKey)
```
Comment (Stebalien) [inline crypto.go:140]: Is this deterministic?

File: crypto.go [hunk 2]
```diff
@@ -95,70 +102,95 @@ func getRemotePubKey(chain []*x509.Certificate) (ic.PubKey, error) {
...
+	if !valid {
+		return nil, errors.New("signature invalid")
+	}
+	return pubKey, nil
 }
 
-func keyToCertificate(sk ic.PrivKey) (crypto.PrivateKey, *x509.Certificate, error) {
-	sn, err := rand.Int(rand.Reader, big.NewInt(1<<62))
+func keyToCertificate(sk ic.PrivKey) (*tls.Certificate, error) {
+	certKey, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
```
Comment (Stebalien) [inline crypto.go:155]: Are there any other curves we can use...?

File: crypto.go [hunk 3]
```diff
@@ -95,70 +102,95 @@ func getRemotePubKey(chain []*x509.Certificate) (ic.PubKey, error) {
...
-		publicKey = &k.PublicKey
-		privateKey = k
-	// TODO: add support for Ed25519
-	default:
-		return nil, nil, errors.New("unsupported key type for TLS")
+	certKeyPub, err := x509.MarshalPKIXPublicKey(certKey.Public())
+	if err != nil {
+		return nil, err
+	}
+	signature, err := sk.Sign(certKeyPub)
```
Comment (Stebalien) [inline crypto.go]: Could we add a prefix? Something like `libp2p-tls-handshake:`? For pubsub, we prefix `libp2p-pubsub:`. I'm probably being paranoid but I'd like to make sure we can't possibly sign a cert in this context and have some other system interpret the signature in a different context.

**Review by Stebalien — COMMENTED** [review]
Nice!

But can we have _lots_ of comments so nobody does something stupid and changes something they shouldn't?

File: crypto.go (hunk 3 above)
Comment (marten-seemann) [reply crypto.go]: Sure. This is needs to be defined in the spec though.

File: crypto.go (hunk 2 above)
Comment (marten-seemann) [reply crypto.go:155]: The Go TLS implementation doesn't support P224, and P256 is curve that produces the smallest signatures. We could also use P384 and P512.

File: crypto.go (hunk 1 above)
Comment (marten-seemann) [reply crypto.go:140]: Yes, it is: https://golang.org/src/crypto/x509/pkcs1.go?s=3633:3686#L138

File: crypto.go (hunk 2 above)
Comment (Kubuxu) [reply crypto.go:155]: Does it support `curve25519`?

Comment (marten-seemann) [reply crypto.go:155]: `keyToCertificate` does (as well as secp256k1), but the Go standard library doesn't. This was one of the reasons we had to invent this new handshake protocol.

Comment (Kubuxu) [reply crypto.go:155]: Ahh, I made a mistake and thought it was `ecdh` not `ecdsa`. (`curve25519` is DH, `ed25519` is signature).

File: crypto.go [hunk 4]
```diff
@@ -95,70 +102,95 @@ func getRemotePubKey(chain []*x509.Certificate) (ic.PubKey, error) {
...
-	certDER, err := x509.CreateCertificate(rand.Reader, tmpl, tmpl, publicKey, privateKey)
+
+	sn, err := rand.Int(rand.Reader, big.NewInt(1<<62))
 	if err != nil {
-		return nil, nil, err
+		return nil, err
+	}
+	tmpl := &x509.Certificate{
+		SerialNumber: sn,
+		NotBefore:    time.Time{},
```
Comment (raulk) [inline crypto.go:186]: Should we add some slack to this timestamp to account for clock skew?

File: crypto.go [hunk 5]
```diff
@@ -1,21 +1,31 @@
...
+	crypto "github.com/libp2p/go-libp2p-crypto"
 	ic "github.com/libp2p/go-libp2p-crypto"
-	pb "github.com/libp2p/go-libp2p-crypto/pb"
 	peer "github.com/libp2p/go-libp2p-peer"
 )
 
-const certValidityPeriod = 180 * 24 * time.Hour
+const certValidityPeriod = 100 * 365 * 24 * time.Hour // ~100 years
+
+var extensionID = getPrefixedExtensionID([]int{1, 1})
```
Comment (raulk) [inline crypto.go:26]: Should we export the extension ID?

File: transport_test.go [hunk 6]
```diff
@@ -37,17 +41,22 @@ var _ = Describe("Transport", func() {
...
+		case 1:
 			fmt.Fprintf(GinkgoWriter, " using an RSA key: ")
-			var err error
 			priv, _, err = ci.GenerateRSAKeyPair(1024, rand.Reader)
-			Expect(err).ToNot(HaveOccurred())
+		case 2:
+			fmt.Fprintf(GinkgoWriter, " using an Ed25519 key: ")
+			priv, _, err = ci.GenerateEd25519Key(rand.Reader)
+		case 3:
+			fmt.Fprintf(GinkgoWriter, " using an Ed25519 key: ")
```
Comment (raulk) [inline transport_test.go]:
```suggestion
			fmt.Fprintf(GinkgoWriter, " using an secp256k1 key: ")
```

**Review by raulk — COMMENTED** [review]
Can't say I'm a crypto expert, but logically LGTM. Minor comments here, and I defer to @Kubuxu and @Stebalien for final approval. Thanks for the great work, @marten-seemann! 🎉

File: crypto.go (hunk 4 above)
Comment (marten-seemann) [reply crypto.go:186]: `time.Time{}` is the smallest possible timestamp.

File: crypto.go (hunk 5 above)
Comment (marten-seemann) [reply crypto.go:26]: I don't see why we would need to do that at the moment, so I'd prefer to keep the API as small as possible.
By the way, what do you think about https://github.com/libp2p/specs/pull/151/files#diff-29603a9a6fd077382b3e7ccb9a175726R39?

File: crypto.go (hunk 4 above)
Comment (raulk) [reply crypto.go:186]: Ah, I read this as `time.Now()` for some reason. Ok!

File: crypto.go (hunk 5 above)
Comment (raulk) [reply crypto.go:26]: See response in https://github.com/libp2p/specs/pull/151/files#r264249948.

**Comment by marten-seemann** [comment]
@Kubuxu, @Stebalien I just merged the TLS spec. Can you please review this PR now?

**Review by Kubuxu — COMMENTED** [review]
SGWM

File: crypto.go [hunk 7]
```diff
@@ -24,22 +37,21 @@ type Identity struct {
...
-			}},
+			MinVersion:               tls.VersionTLS13,
+			PreferServerCipherSuites: preferServerCipherSuites(),
+			InsecureSkipVerify:       true, // This is not insecure here. We will verify the cert chain ourselves.
+			ClientAuth:               tls.RequireAnyClientCert,
+			Certificates:             []tls.Certificate{*cert},
 			VerifyPeerCertificate: func(_ [][]byte, _ [][]*x509.Certificate) error {
 				panic("tls config not specialized for peer")
 			},
+			SessionTicketsDisabled: true,
```
Comment (Stebalien) [inline crypto.go:54]: Any reason for this?

**Review by Stebalien — APPROVED** [review]

File: crypto.go (hunk 7 above)
Comment (marten-seemann) [reply crypto.go:54]: When resuming a session using session tickets, no certificate chain is presented, and the callbacks needed to verify the peer identity would not be called.

We can add session ticket support later (and save the peer ID alongside the session ticket in the session ticket cache).

**Pull request merged**